### PR TITLE
samples: fix the metairq_dispatch sample sporadically failed on ehl_crb

### LINKED ISSUE #40889: samples: samples/kernel/metairq_dispatch failed on acrn_ehl_crb
**Describe the bug**
The sample samples/kernel/metairq_dispatch runs failed on acrn_ehl_crb

Please also mention any information which could help others to understand
the problem you're facing:
 - What target platform are you using?
   acrn_ehl_crb
 - What have you tried to diagnose or workaround this issue?
    It seems like that local variable ( in function calc_stats( ) ) was optimized incorrectly. 

**To Reproduce**
1. source zephyr-env.sh
2. twister -p acrn_ehl_crb --device-testing --device-serial-pty="/opt/remotehw/acrn-test-pty.exp,ehlsku11" --west-flash="/opt/remotehw/remotehw-x86-acrn.sh,ehlsku11" -vv -T samples/kernel/metairq_dispatch/
3. See error

**Expected behavior**
The sample run successfully and The console output should show a complete statistical message.

**Impact**
No serious impact. 

**Logs and console output**
```
DEBUG   - DEVICE: ACRN:\>vm_console 0
DEBUG   - DEVICE:
DEBUG   - DEVICE: ----- Entering VM 0 Shell -----
DEBUG   - DEVICE: *** Booting Zephyr OS build v2.7.99-2074-gfd089b361d8a  ***
DEBUG   - DEVICE: I: Starting Thread0 at priority -2
DEBUG   - DEVICE: I: Starting Thread1 at priority -1
DEBUG   - DEVICE: I: Starting Thread2 at priority 0
DEBUG   - DEVICE: I: Starting Thread3 at priority 1
DEBUG   - DEVICE: I: M0 TI: M2 T1 mirq 56424 disp 83980 proc 846417 real 846644
DEBUG   - DEVICE: I: M5 T2 mirq 5716I: M6 T1 mirq 56386 disp 83924 proc 188953 real 190768
DEBUG   - DEVICE: I: M7 T1 mirq 57124 disp 785668 proc 114430 real 115970
DEBUG   - DEVICE: 2 I: M9 T1 mirq 57190 disp 84594 proc 867500 real 868808
DEBUG   - DEVICE: I: M12 T0 mirq 56380 disp 91980 proc 672098 real 672468
DEBUG   - DEVICE: I: M14 T1 mirq 56530 disp 1028312 proc 866741 real 868270
DEBUG   - DEVICE: I: M16 T0 mirq 56546 disp 1017768 proc 717969 real 756798
DEBUG   - DEVICE: I: M19 T1 mirq 57092 disp 314958 proc 690395 real 691756
DEBUG   - DEVICE: I: M20 T1 mirq 57148 disp 2267322 proc 385504 real 386634
DEBUG   - DEVICE: I: M22 T1 mirq 71104 disp 2862902 proc 118718 real 119352
DEBUG   - DEVICE: I: M25 T1 mirq 56402 disp 1206860 proc 543519 real 561226
DEBUG   - DEVICE: I: M26 T1 mirq 57144 disp 2633466 proc 634371 real 635170
DEBUG   - DEVICE: I: M29 T1 mirq 57026 disp 2570214 proc 144982 real 154106
DEBUG   - DEVICE: I: M30 T0 mirq 56370 disp 2948338 proc 514620 real 515072
DEBUG   - DEVICE: I: M31 T0 mirq 56514 disp 3789468 proc 867840 real 868752
DEBUG   - DEVICE: I: M33 T1 mirq 56484 disp 4126058 proc 786118 real 787146
DEBUG   - DEVICE: I: M35 T1 mirq 60204 disp 3712082 proc 499509 real 501218
DEBUG   - DEVICE: I: M38 T0 mirq 65092 disp 2034580 proc 421149 real 422474
DEBUG   - DEVICE: I: M40 T0 mirq 44324 disp 2505880 proc 831208 real 832834
DEBUG   - DEVICE: disp 316524 proc 715912 real 717534
DEBUG   - DEVICE: I: M21 T2 mirq 57018 disp 26253534 proc 713039 real 713698
DEBUG   - DEVICE: I: M24 T2 mirq 57126 disp 25223140 proc 557922 real 558824
DEBUG   - DEVICE: I: M34 T2 mirq 56444 disp 17181176 proc 101896 real 102952
DEBUG   - DEVICE: I: M37 T2 mirq 63518 disp 14772128 proc 6528 real 7084
DEBUG   - DEVICE: 3 mirq 62934 disp 1848870 proc 233453 real 233920
DEBUG   - DEVICE: I: M1 T3 mirq 58130 disp 53738516 proc 155711 real 156450
DEBUG   - DEVICE: I: M3 T3 mirq 57842 disp 52893486 proc 387136 real 387982
DEBUG   - DEVICE: I: M4 T3 mirq 57526 disp 54362664 proc 272484 real 273094
DEBUG   - DEVICE: I: M8 T3 mirq 61024 disp 51353502 proc 616519 real 617626
DEBUG   - DEVICE: I: M10 T3 mirq 56002 disp 52290818 proc 215855 real 216560
DEBUG   - DEVICE: I: M11 T3 mirq 57094 disp 52660366 proc 627772 real 629636
DEBUG   - DEVICE: I: M13 T3 mirq 57160 disp 52864904 proc 323243 real 324562
DEBUG   - DEVICE: I: M15 T3 mirq 56816 disp 52798520 proc 384262 real 384636
DEBUG   - DEVICE: I: M17 T3 mirq 57094 disp 52389796 proc 96409 real 97278
DEBUG   - DEVICE: I: M18 T3 mirq 56298 disp 53553218 proc 636733 real 638310
DEBUG   - DEVICE: I: M23 T3 mirq 56604 disp 51505714 proc 84139 real 85370
DEBUG   - DEVICE: I: M27 T3 mirq 57120 disp 49039226 proc 358939 real 359164
DEBUG   - DEVICE: I: M28 T3 mirq 56566 disp 50119392 proc 332149 real 333634
DEBUG   - DEVICE: I: M32 T3 mirq 57408 disp 47759630 proc 180376 real 181064
DEBUG   - DEVICE: I: M36 T3 mirq 67948 disp 44096066 proc 312837 real 314280
DEBUG   - DEVICE: I: M39 T3 mirq 64366 disp 41729202 proc 133903 real 134678
DEBUG   - DEVICE: I:         ---------- Latency (cyc) ----------
DEBUG   - DEVICE: I:             Best    Worst     Mean    Stdev
DEBUG   - DEVICE: I: MetaIRQ    44324    71104    58065     4025
DEBUG   - DEVICE: E: Page fault at address 0x129000 (error code 0x0)
DEBUG   - DEVICE: E: Linear address not present in page tables
DEBUG   - DEVICE: E: Access violation: supervisor thread not allowed to read
DEBUG   - DEVICE: E: PTE: not present
DEBUG   - DEVICE: E: RAX: 0x000001a4b82fa8af RBX: 0x0000000000000000 RCX: 0x0000000000115fac RDX: 0x0000000000115fb0
DEBUG   - DEVICE: E: RSI: 0x000000000000fb6e RDI: 0x0000000000129000 RBP: 0x0000000000115f90 RSP: 0x0000000000115f80
DEBUG   - DEVICE: E:  R8: 0x0000000000115fb4  R9: 0x0000000000115fb8 R10: 0x000000000010c0ac R11: 0x00000000000073d5
DEBUG   - DEVICE: E: R12: 0x0000000000000003 R13: 0x0000000000000000 R14: 0x000000000010c000 R15: 0x0000000000115fb0
DEBUG   - DEVICE: E: RSP: 0x0000000000115f80 RFLAGS: 0x0000000000010206 CS: 0x0018 CR3: 0x0000000000128000
DEBUG   - DEVICE: E: RIP: 0x000000000010005b
DEBUG   - DEVICE: E: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0
DEBUG   - DEVICE: E: Current thread: 0x10d360 (unknown)
DEBUG   - DEVICE: E: Halting system
DEBUG   - Timed out while monitoring serial output on acrn_ehl_crb
DEBUG   - Process /opt/remotehw/acrn-test-pty.exp,ehlsku11 terminated outs: None errs None
DEBUG   - run status: acrn_ehl_crb/samples/kernel/metairq_dispatch/sample.kernel.metairq_dispatch timeout
INFO    - 1/1 acrn_ehl_crb              samples/kernel/metairq_dispatch/sample.kernel.metairq_dispatch FAILED Timeout (device 78.483s)
ERROR   - see: /home/emai/work/zephyrproject/zephyr/twister-out/acrn_ehl_crb/samples/kernel/metairq_dispatch/sample.kernel.metairq_dispatch/handler.log
```

**Environment (please complete the following information):**
 - OS: Linux
 - Toolchain: Zephyr SDK 0.13.2
 - Commit SHA or Version used

**Additional context**
N/A

### LINKED ISSUE #42874: ehl_crb: samples/kernel/metairq_dispatch fails when it is run multiple times
**Describe the bug**
The sample samples/kernel/metairq_dispatch runs failed on ehl_crb

**To Reproduce**

1. source zephyr-env.sh
2. twister -W -p ehl_crb --device-testing --device-serial /dev/ttyUSB0 -T samples/kernel/metairq_dispatch/ --west-flash=/home/ztest/EHL_X86_PXE.sh -j 1 -vvv
3. Sample passes when run for the first time and fails when we re-run again.

**Expected behavior**
The sample should run successfully and console output should show a complete statistical message.

**Impact**
No serious impact.

**Logs and console output**

```
[2J[01;01H[=3h[2J[01;01H[2J[01;01H[=3h[2J[01;01H
Press [Enter]  to directly boot.
Press [F2]     to enter setup and select boot options.
Press [F7]     to show boot menu options.
Press [F12]    to PXE boot.
Press [Pause]  to pause and <any key> to continue booting.
**  WARNING: Test Key is used.  **
>>Start PXE over IPv4.
Station IP address is 192.165.1.99
Server IP address is 192.165.1.1
NBP filename is zephyr.efi
NBP filesize is 2191459 Bytes
Downloading NBP file...
NBP file downloaded successfully.
*** Zephyr EFI Loader ***
Zeroing 4802248 bytes of memory at 0x10c000
Copying 24576 data bytes to 0x1000 from image offset 
Copying 49152 data bytes to 0x100000 from image offset 24576
Copying 2111776 data bytes to 0x5a06e0 from image offset 73728
Jumping to Entry Point: 0x10db (48 31 c0 48 31 d2 48)
*** Booting Zephyr OS build v3.0.0-rc3-21-gccf51e3cd3ae  ***
I: Starting Thread0 at priority -2
I: Starting Thread1 at priority -1
I:I: M0 T0 mirq 1751 disp 2243 proc 3276 real 4123
I: M9 T0 mirq 3334 disp 37516 proc 6785 real 7387
I: M12 T0 mirq 2096 disp 109552 proc 3680 real 5779
I: M13 T0 mirq 2097 disp 203266 proc 6046 real 6464
I: M14 T0 mirq 2072 disp 297287 proc 5458 real 5758
I: M22 T0 mirq 2096 disp 329534 proc 2382 real 2857
I: M24 T0 mirq 2097 disp 409892 proc 2705 real 3179
I: M38 T0 mirq 2097 disp 385401 proc 5284 real 5762
I: M40 T0 mirq 1129 disp 464574 proc 5872 real 6404
I: M2 T1 mirq 1451 disp 901655 proc 579 real 922
I: M4 T1 mirq 2096 disp 972726 proc 3583 real 4147
I: M5 T1 mirq 2073 disp 1057759 proc 5061 real 5437
I: M7 T1 mirq 2096 disp 1147783 proc 5215 real 5760
I: M8 T1 mirq 3656 disp 1245657 proc 8168 real 8663
I: M10 T1 mirq 2096 disp 1341470 proc 902 real 1244
I: M11 T1 mirq 2073 disp 1425645 proc 6981 real 7370
I: M15 T1 mirq 2097 disp 1489131 proc 2949 real 3501
I: M16 T1 mirq 1452 disp 1582959 proc 7549 real 8019
I: M17 T1 mirq 1429 disp 1683112 proc 8958 real 9307
I: M23 T1 mirq 2096 disp 1739625 proc 5312 real 5758
I: M34 T1 mirq 2096 disp 1747800 proc 6839 real 7373
I: M35 T1 mirq 1451 disp 1843557 proc 5877 real 6405
I: M36 T1 mirq 2096 disp 1938155 proc 3400 real 3823
Starting Thread2 at priority 0
I: M3 T2 mirq 2096 disp 2381496 proc 2984 real 3502
I: M19 T2 mirq 2095 disp 2344522 proc 1122 real 1567
I: M20 T2 mirq 2097 disp 2439927 proc 1668 real 2211
I: M26 T2 mirq 2096 disp 2490641 proc 5637 real 6082
I: M29 T2 mirq 2095 disp 2570126 proc 6946 real 7372
I: M33 T2 mirq 1452 disp 2639712 proc 1971 real 2533
I: M37 T2 mirq 2096 disp 2709274 proc 5073 real 5436
I: Starting Thread3 at priority 1
I: M1 T3 mirq 2097 disp 3179895 proc 6806 real 7370
I: M6 T3 mirq 2098 disp 3234666 proc 8244 real 8659
I: M18 T3 mirq 2096 disp 3240251 proc 91 real 599
I: M21 T3 mirq 2074 disp 3315055 proc 4948 real 5435
I: M25 T3 mirq 2096 disp 3386481 proc 4966 real 5436
I: M27 T3 mirq 2095 disp 3469219 proc 1019 real 1567
I: M28 T3 mirq 2095 disp 3558985 proc 553 real 921
I: M30 T3 mirq 2096 disp 3645039 proc 2788 real 3178
I: M31 T3 mirq 2095 disp 3738534 proc 8949 real 9305
I: M32 T3 mirq 1451 disp 3834344 proc 9185 real 9628
I: M39 T3 mirq 2073 disp 3868267 proc 4658 real 5136
```
Environment (please complete the following information):

OS: Linux
Toolchain: Zephyr SDK 0.13.2
Commit SHA: ccf51e3cd3ae

### PR DESCRIPTION
**Pull request by enjiamai**
The metairq sample sometimes failed due to stuck because it has a
potential race when recording the latencies. The previous solution
checks whether the stats.mirq greater than MAX_EVENT. But stats.mirq
might also have a race here. So we change to check the msg sequence,
its correctness ensured by the kernel's message queue.

And also remove the spinlock in previous fix because tests only use
1 CPU, no need to add this. It also improves the fix for #40889,
to prevent test failed sporadically on acrn_ehl_crb and ehl_crb.	

Fix #42874.

Signed-off-by: Enjia Mai <enjia.mai@intel.com>

### CODE REVIEW

File: samples/kernel/metairq_dispatch/src/main.c [hunk 1]
```diff
@@ -191,7 +197,7 @@ static void thread_fn(void *p1, void *p2, void *p3)
 		 * and get queued.
 		 */
 		while (k_cycle_get_32() - start < m.proc_cyc) {
-			for (volatile int i = 0; i < 100; i++) {
+			for (volatile int i = 0; i < SPIN_TIMES; i++) {
```
Comment (andyross) [inline samples/kernel/metairq_dispatch/src/main.c:200]: This doesn't increase the time spent waiting, it just decreases the frequency with which the timer is polled.  We wait for `m.proc_cyc` in either case.

If this is fixing something, the root cause has to be elsewhere.

File: samples/kernel/metairq_dispatch/src/main.c [hunk 2]
```diff
@@ -120,13 +118,6 @@ static void record_latencies(struct msg *m, uint32_t latency)
 		return;
 	}
 
-	k_spinlock_key_t key = k_spin_lock(&lock);
```
Comment (andyross) [inline samples/kernel/metairq_dispatch/src/main.c:123]: I'm confused here.  I'm just now noticing the earlier patch, but... it looks correct to me.  This is a test of preemptible thread behavior, so locking of some form between the threads is probably what we want.

**Review by andyross — CHANGES_REQUESTED** [review]
I'm a little suspicious here.  Lacking a root cause, I'd say that the earlier fix (synchronize the record_latencies() function) was correct.  The new one is very confusing, and IMHO probably points to something like a latency bug with the timer driver.

File: samples/kernel/metairq_dispatch/src/main.c (hunk 2 above)
Comment (andyross) [reply samples/kernel/metairq_dispatch/src/main.c:123]: But to be clear: it doesn't have to be a spinlock.  There's no ISR context and (as you point out) we're running with MP_NUM_CPUS=1, so anything will do: a k_mutex() or k_sem() would be fine, and (as long as we're never SMP) so would sched_un/lock().

File: samples/kernel/metairq_dispatch/src/main.c (hunk 1 above)
Comment (enjiamai) [reply samples/kernel/metairq_dispatch/src/main.c]: You are absolutely right. This root cause is due to race on record_latencies subroutine, add timeout here is not a right way to go. Correct it. thanks for pointing this out.

File: samples/kernel/metairq_dispatch/src/main.c (hunk 2 above)
Comment (enjiamai) [reply samples/kernel/metairq_dispatch/src/main.c:123]: Yes, @andyross , in previous patch I did not aware of this is a 1 CPU scenario, like you pointing out, it only does irq_lock here, and no ISR context. (Actually I tried it doesn't work at all to fix this issue).  So the spinlock should be removed here.

And I try on another way to fix this, by just checking if the message sequence number > MAX_EVENT) when entering the record_latencies subroutine  (in previously I check the stats.mirq but it still failed sometimes due to this shared variable can also have a race. 

Look like it works well and no need to protect a section of code. And I tested it on both ehl_crb and acrn_ehl_crb then the test pass more than 20 times. Actually this failure can be reproduced within 5~8 times previously.

Do you think is this solution ok? Thank you!

**Review by peter-mitsis — APPROVED** [review]

File: samples/kernel/metairq_dispatch/src/main.c [hunk 3]
```diff
@@ -120,10 +118,11 @@ static void record_latencies(struct msg *m, uint32_t latency)
...
-
-	if (stats.num_mirq >= MAX_EVENTS) {
-		k_spin_unlock(&lock, key);
+	/* It might be a potential race condition in this subroutine.
+	 * We check if the msg seqence is reaching the MAX EVENT first.
+	 * To prevent the coming incorrect changes of the array.
+	 */
```
Comment (andyross) [inline samples/kernel/metairq_dispatch/src/main.c:124]: I'd still prefer to see a root cause here, and don't understand why the previous fix with the spinlock wasn't correct.  But this looks safe as I read it.

**Review by andyross — APPROVED** [review]

**Pull request merged**